JOSE_paper/paper.md

```diff
@@ -0,0 +1,254 @@
+- Code formatting for consistent readable layout,
+- Naming principles for readable, re-useable code,
+- Linting and static analysis of code,
+- Writing docstrings and comments, and
+- Other general principles including magic numbers, avoiding hard-coded values, etc.
+
+This makes it easy to remove sections to tailor the course, or for additional topics to be added in future.
+
+## Exercises
+
+The slides are interspersed with a series of exercises, each one applying techniques
+that have just been taught.
+
+We set up a scenario where users have been provided a working, but poorly written piece
+of code that they are now expected to re-use in their work.
+As they progress through the exercises they apply their new skills to transform this
+into a much more readable and understandable piece of code.
+
+The script we use is based on some of the code and data available in @Irving2019.
+This was chosen as it features typical applications we expect users to encounter
+(reading from a dataset, processing, and plotting).
+The course was originally designed to be taught to geoscientists, and then at a climate
+science summer school, and we wanted participants to be focussed on the changes to the
+code rather than worrying about the application or semantics.
+
+Exercises roughly match the slide sections above:
+
+- 01: Baseline code to examine, install dependencies, and run.\
+  Users can familiarise themselves with what the code does, see why listing dependencies
+  with a quick setup is useful, and make early observations about how the code is
+  difficult to understand/use.
+- 02: Apply a formatter to standardise code.\
+  We use the black python formatter [@LangaBlack]
+- 03: Improve code clarity with naming and source changes.
+- 04: Linting and static analysis of code.
+- 05: Writing docstrings and best use of comments.
+- 06: General techniques for better code (magic numbers, string formatting).
+- 00: The end point of the workshop - an improved version of the code in exercise 01.
+
+Each exercises builds upon the results of the previous one, allowing participants to
+apply what they are learning in a continuous fashion to a code.
+However, each exercise has a dedicated subdirectory meaning that success in a previous
+exercise is not a pre-requisite for continuing to the next one.
+Two additional benefits to this approach are that the starting point for a subsequent
+exercise can be viewed as the solution to the current one, allowing participants to
+compare their work to an 'ideal' solution, and that exercises can easily be removed to
+tailor the course to different emphasis or time constraints.
+
+
+
+# Content Delivery
+
+The course has been designed to be flexible in terms of delivery, allowing
+it to be adapted to and reused in various environments.
+We have adopted a modular design for the course material -- using Quarto for the slides
+with each section in a separate file, and each associated exercise in a separate
+directory.
```

Comment:
Mention that the same is possible, for example, with Latex?

Reply:
My reason to mention Latex was that we mention doing this with quarto, and people might say "oh, but that is not a unique thing about quarto". But maybe it does not need to be mentioned.

Reply:
Hm, I had another read through it, and maybe we just leave it. The tool used for rendering the material should not be the main point here.

Reply:
I have amended this to simply say that we have a modular structure to the slides :)